### PR TITLE
FCM Delivery Method Updates

### LINKED ISSUE #378: [BUG] FCM Delivery Method errors when config[:json] is a method
### Describe the Bug:
at `/lib/noticed/delivery_methods/fcm.rb:17` 
the line
``` ruby
notification.instance_exec(device_token, &config[:json]))
```
in the `send_notification` method

Throws ` NoMethodError (undefined method 'android_format' for an instance of String):`

when `fcm.config.json = :android_format`

and `android_format` is a method that takes `device_token` as an argument and returns JSON

### To Reproduce:

Configure FCM Delivery Method
``` ruby
deliver_by :fcm do |config|
  config.credentials = Rails.root.join("config/certs/fcm.json")
  config.device_tokens = [ "eUlu_npTtR565sEADHeGXV:APA91bGgm2rssXgBdhjt9DDTZJShjnr_8OpxUB29zQQ4CvFCAl1exk9gYXluBJvj6q7EdMbO72A5E-N_LCHiIAGOOTiHqNo-iEwlckvEl2yfurbLCwRg0w84GqOf19pHgwaGd_1nZLF4" ]
  config.json = :android_format
end

def android_format(device_token)
  {
    token: device_token,
    notification: {
      title: "Test Title",
      body: "Test body"
    }
  }
end
```
Trigger a notification. 

### Expected Behavior

Notification to send correctly. 

### Actual Behavior

``` bash
2024-01-23T11:25:31.341Z pid=232233 tid=5im9 class=Noticed::DeliveryMethods::Fcm jid=26dbb4c80f7adef7db91de3c ERROR: Error performing Noticed::DeliveryMethods::Fcm (Job ID: 27faecc0-1464-42c0-b6f1-4888451ae0cd) from Sidekiq(default) in 332625.98ms: NoMethodError (undefined method `android_format' for an instance of String):
/home/phil/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/noticed-2.0.3/lib/noticed/delivery_methods/fcm.rb:18:in `instance_exec'
/home/phil/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/noticed-2.0.3/lib/noticed/delivery_methods/fcm.rb:18:in `send_notification'
/home/phil/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/noticed-2.0.3/lib/noticed/delivery_methods/fcm.rb:10:in `block in deliver'
/home/phil/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/noticed-2.0.3/lib/noticed/delivery_methods/fcm.rb:9:in `each'
/home/phil/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/noticed-2.0.3/lib/noticed/delivery_methods/fcm.rb:9:in `deliver'
/home/phil/.rbenv/versions/3.3.0/lib/ruby/gems/3.3.0/gems/noticed-2.0.3/lib/noticed/delivery_method.rb:21:in `perform'
```

### Environment:
- Noticed gem version: 2.0.3
- Ruby version: 3.3.0
- Rails version: 7.1.3
- Operating System: Ubuntu 22.04

### Possible Fix:

Copy pattern from ios.md and implement a `format_notification` method.

Note: I'm not really a fan of the nested conditionals here

``` ruby
def format_notification(device_token)
  if (method = config[:json])
    if method.is_a?(Symbol) && event.respond_to?(method)
      event.send(method, device_token)
    else
      notification.instance_exec(device_token, &method)
    end
  else
    raise ArgumentError, "No json for fcm delivery. Add the 'json' option in 'deliver_by :fcm'."
  end
end
```
send notification method would then become

``` ruby
def send_notification(device_token)
  post_request("https://fcm.googleapis.com/v1/projects/#{credentials[:project_id]}/messages:send",
    headers: {authorization: "Bearer #{access_token}"},
    json: format_notification(device_token))
rescue Noticed::ResponseUnsuccessful => exception
  if exception.response.code == "404" && config[:invalid_token]
    notification.instance_exec(device_token, &config[:invalid_token])
  else
    raise
  end
end
```

This should mean that existing behaviour is not modified but adds the enhancement that `config[:json]` can be a symbol representation of a method.

Happy to put this up for PR if you're happy with the nested conditional.

### Checklist:

- [x] I have searched for similar issues and couldn't find any
- [x] I have checked the documentation for relevant information
- [x] I have included all the required information

### PR DESCRIPTION
**Pull request by phil-6**
## Summary
Updates `lib/noticed/delivery_methods/fcm.rb` to restore `message {}` wrapper around JSON body and restore functionality to allow `config[:json]` to be a symbol representation of a method.

## Related Issue
Fixes #378 

## Testing
All Tests are passing
Tested with our live application to confirm functionality with both the JSON as a lambda and a symbol

## Checklist

- [x] Code follows the project's coding standards
- [x] Tests have been added or updated to cover the changes
- [x] Documentation has been updated (if applicable)
- [x] All existing tests pass
- [x] Conforms to the contributing guidelines

---

As per my other PR, I'm fairly new to the open source contribution game so any feedback is appreciated! 

I'm not sure of the `format_notification` method. Generally I prefer early returns over nested conditionals, but I'm not certain about this one!

### CODE REVIEW

**Comment by excid3** [comment]
Forgot about this. I believe my thinking was by removing the message wrapper, we are more compatible with the FCM docs and any future changes. Since this would be a breaking change, we'd ideally do this in v2's release and update the examples.

I'm still leaning towards leaving the message wrapper out for long-term support and maintenance. That said, FCM probably isn't going to change a ton? 

For example, we do not support this `validate_only` option currently and would have to implement a workaround: https://firebase.google.com/docs/reference/fcm/rest/v1/projects.messages/send
The other benefit is you can copy the examples from FCM docs/tutorials and use that without removing the message wrapper to translate to Noticed configs.

Thoughts?

**Comment by phil-6** [comment]
I think either way is fine, as long as it's consistent. 

From what I've seen FCM seems to be very good with preserving old functionality - you can still use the OG Server Tokens even if you can't generate new ones - so I suspect that any change to the JSON structure would offer the existing structure as an option for an extended period. BUT it is Google so who knows!

Would you like me to revert that change and the docs updates?

**Comment by excid3** [comment]
Yeah, let's drop the `message` wrapper and leave it more flexible. If Google ever changes things, we'll be in a good position to already support it.

File: lib/noticed/delivery_methods/fcm.rb [hunk 1]
```diff
@@ -23,6 +23,16 @@ def send_notification(device_token)
         end
       end
 
+      def format_notification(device_token)
+        raise ArgumentError, "No json for fcm delivery. Add the 'json' option in 'deliver_by :fcm'." unless (method = config[:json])
```
Comment (excid3) [inline lib/noticed/delivery_methods/fcm.rb:27]: Is this redundant since we already have `required_option :json`?

Comment (phil-6) [reply lib/noticed/delivery_methods/fcm.rb]: Yep. I missed that. Thanks.

Removing it tidies up that method nicely too

File: lib/noticed/delivery_methods/fcm.rb [hunk 2]
```diff
@@ -23,6 +23,15 @@ def send_notification(device_token)
         end
       end
 
+      def format_notification(device_token)
```
Comment (excid3) [inline lib/noticed/delivery_methods/fcm.rb:26]: This is a good idea. 👍

Comment (phil-6) [reply lib/noticed/delivery_methods/fcm.rb:26]: Thank you! It's your idea originally :grinning:

Comment (excid3) [reply lib/noticed/delivery_methods/fcm.rb:26]: LOL

**Pull request merged**